Teradata Vantage - Added SSL support

SatishChGit opened the pull request:
Modified database size from 1GB to 120 MB for schema creating from Airbyte.

## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

## How
*Describe the solution*

## Recommended reading order
1. `[TeradataDestination.java](https://github.com/airbytehq/airbyte/compare/master...Teradata:airbyte:Teradata_SSL_Support?expand=1#diff-81950dec5336116d6abe1fb35b027346cb6538ed93f6d8f727b559f2fae1af46)`
2. `airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataSqlOperations.java`
3. 'airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json'

## 🚨 User Impact 🚨
No breaking changes. User can load data from any airbyte supported source to teradata through airbyte.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

![SSL_UnitResult](https://user-images.githubusercontent.com/23031170/214181163-a25ed0e5-64a5-4819-8182-b54bce58b815.png)


</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

![SSL_AcceptanceResult](https://user-images.githubusercontent.com/23031170/214181181-6583adb1-f6bf-4f97-9f5a-63b86fa794c0.png)

</details>

grishick commented on airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java:
```diff
@@ -12,57 +12,108 @@
+    @Override
+    protected Map<String, String> getDefaultConnectionProperties(final JsonNode config) {
+        final Map<String, String> additionalParameters = new HashMap<>();
+        LOGGER.info("teradata config obj : " + config);
```
please remove this log statement to avoid leaking any sensitive config information

SatishChGit replied:
Removed this log statement.

grishick commented on airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:
```diff
@@ -0,0 +1,162 @@
+
+  public JsonNode getStaticConfig() throws Exception {
+    final JsonNode config = Jsons.deserialize(Files.readString(Paths.get("secrets/sslconfig.json")));
+    ((ObjectNode) config).put("ssl", true);
```
why not have `ssl` and `sslmode` specified in `secrets/sslconfig.json` instead of adding them programmatically.

grishick replied:
also, `spec.json` expects `ssl_mode` but here you are adding `sslmode`

SatishChGit replied:
Removed providing ssl options from this class and will provide latest config json file.

SatishChGit replied:
Yes, spec.json expects ssl_mode. This testcase was got succeeded when ssl is true, if no ssl_mode specified with any value like require, prefer.., it is taking require as default value. 
![image](https://user-images.githubusercontent.com/23031170/214711346-d1969e0e-a59f-477a-99f1-f362a9c0cb5a.png)

grishick commented on airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json:
```diff
@@ -21,28 +21,153 @@
         "type": "string",
-        "examples": ["public"],
-        "default": "public",
+        "examples": ["airbyte_td"],
```
this change seems unnecessary

SatishChGit replied:
public is keyword in Teradata system, and it will not allow to create schema with public as default value.  Changed it to airbyte_td

grishick commented on airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:
```diff
@@ -0,0 +1,162 @@
+
+public class TeradataDestinationSSLAcceptanceTest extends JdbcDestinationAcceptanceTest {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestinationAcceptanceTest.class);
```
this member is not used, please remove

SatishChGit replied:
Removed.

grishick commented on airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationAcceptanceTest.java:
```diff
@@ -95,7 +95,7 @@ private List<JsonNode> retrieveRecordsFromTable(final String tableName, final St
     final String schemaName = Strings.addRandomSuffix("integration_test_teradata", "_", 5);
     final String createSchemaQuery = String
-        .format(String.format("CREATE DATABASE \"%s\" AS PERM = 1e9 SKEW = 10 PERCENT", schemaName));
+        .format(String.format("CREATE DATABASE \"%s\" AS PERMANENT = 60e6, SPOOL = 60e6 SKEW = 10 PERCENT", schemaName));
```
why is this change relevant to this PR?

SatishChGit replied:
1GB is not required for running integration tests. With this huge memory setting for each database creating while running integration tests, space issues are coming and test cases getting failed. So changed this to limited memory to make integration tests gets success.

grishick commented on airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:
```diff
@@ -0,0 +1,162 @@
+  private JdbcDatabase database;
+  private DataSource dataSource;
+  private TeradataDestination destination = new TeradataDestination();
+  private final JdbcSourceOperations sourceOperations = JdbcUtils.getDefaultSourceOperations();
```
`sourceOperations` is not used anywhere in this class, so it should be removed

SatishChGit replied:
Removed.

grishick commented:
This PR is looking better. Noting the issues we've discussed over slack to keep track:
 - need to make DB storage footprint configurable via spec instead of hardcoded
 - need to drop test DBs in `tearDown` method in `TeradataDestinationAcceptanceTest`
 - integration tests should be conservative when specifying storage footprint for test DBs (this would help avoid running out of space on the test instance)
 - avoid modifying static config in test code if possible and use config files instead. I.e. `sslconfig.json` should have SSL config enabled

SatishChGit commented:
> This PR is looking better. Noting the issues we've discussed over slack to keep track:
> 
> * need to make DB storage footprint configurable via spec instead of hardcoded
   I tried to make use of spec to specify storage footprint but couple of testcases are failing. Will work on this in different PR request as currently taking time to investigate on these failures. Will create JIRA from ourside and work in coming sprints. 
> * need to drop test DBs in `tearDown` method in `TeradataDestinationAcceptanceTest`
     Implemented dropping test DBs through tearDown.
> * integration tests should be conservative when specifying storage footprint for test DBs (this would help avoid running out of space on the test instance)
   I tried to make use of spec to specify storage footprint but couple of testcases are failing. Will work on this in different PR request as currently taking time to investigate on these failures. Will create JIRA from ourside and work in coming sprints. 
> * avoid modifying static config in test code if possible and use config files instead. I.e. `sslconfig.json` should have SSL config enabled
 Modified as per this request

SatishChGit commented:
Integration tests results after all review comments addressed.
![image](https://user-images.githubusercontent.com/23031170/216249224-5d3a15cb-bcdc-482f-8e7b-709fd9cc0af8.png)

Unit tests results after all review comments addressed.
![image](https://user-images.githubusercontent.com/23031170/216249371-060c8909-e7a3-4b6b-ac2a-ce064706338e.png)

grishick commented on airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:
```diff
@@ -0,0 +1,17 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestinationSSLAcceptanceTest.class);
+
+  protected  String getConfigFileName() {
```
this method is not being called anywhere as far as I can see, which means that this test runs w/o SSL

grishick commented on airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java:
```diff
@@ -12,57 +12,107 @@
-
-    if (config.has(JdbcUtils.PASSWORD_KEY)) {
-      configBuilder.put(JdbcUtils.PASSWORD_KEY, config.get(JdbcUtils.PASSWORD_KEY).asText());
+    private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestination.class);
```
LOGGER appears to be unused

grishick commented:
/test connector=connectors/destination-teradata

grishick commented:
note: I ran tests locally and they passed

grishick commented:
@SatishChGit in addition to the two comments I left earlier, could you also rebase your feature branch on latest master or merge latest master into your feature branch - we have some test and instrumentation changes that are not in your feature branch and that's causing test failures.

SatishChGit commented:
> @SatishChGit in addition to the two comments I left earlier, could you also rebase your feature branch on latest master or merge latest master into your feature branch - we have some test and instrumentation changes that are not in your feature branch and that's causing test failures.

Okay Sure Greg. Will do and run tests

grishick commented:
I think we can merge this as an alpha connector. Before we can add it to Cloud, we'd need to:
1. integrate spinning up/down of teradata instance into Before/After hooks (we have a discussion in Slack going about this)
2. make sure that only integration tests require a running instance (filed https://github.com/airbytehq/airbyte/issues/23121 to track)
3. some cleanup: https://github.com/airbytehq/airbyte/issues/23122

SatishChGit commented:
> I think we can merge this as an alpha connector. Before we can add it to Cloud, we'd need to:
> 
> 1. integrate spinning up/down of teradata instance into Before/After hooks (we have a discussion in Slack going about this)
> 2. make sure that only integration tests require a running instance (filed [`TeradataDestinationTest` should not require a running Teradata instance #23121](https://github.com/airbytehq/airbyte/issues/23121) to track)
> 3. some cleanup: [Unused members in `TeradataDestinationTest` #23122](https://github.com/airbytehq/airbyte/issues/23122)

Sure Greg. After merging master to this branch, couple of test cases are failing in my local environment. I am checking on them and will update once these run successfully.
